### PR TITLE
Gather user projects and Testing for projects route

### PR DESCRIPTION
**Pull request by ruv-prog-so**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -25,6 +25,7 @@
     "serve-favicon": "~2.3.0"
   },
   "devDependencies": {
+    "ajv": "^4.1.7",
```
Comment (RickyV33) [inline package.json]: We'll be using AJV for validation in our code as well so this can be a Dependency instead

File: routes/index.js [hunk 2]
```diff
@@ -1,5 +1,6 @@
 let express = require('express');
 let auth = require('../lib/auth');
+let proj = require('../lib/projects');
```
Comment (chances) [inline routes/index.js]: Use the full, plural word

**Comment by chances** [comment]
Convert `schema/projectList.js` to a JSON file, `lib/schema/projectList.json`

File: test/routes/projects.js [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+// For routes test
+let ejs = require('ejs');
+let read = require('fs').readFileSync;
+let join = require('path').join;
+
+chai.use(chaiHttp);
+let mockProjects = require('./mockProjects.json');
+
+let credentialFixtureCases = [
```
Comment (chances) [inline test/routes/projects.js]: Rename this to `projectFixtureCases`

File: views/projects.ejs [hunk 4]
```diff
@@ -13,4 +13,4 @@
 <!-- TODO: On a post to this page, don't change the title of the page to the error message -->
 <script src="js/project.js"></script>
 </body>
-</html>
+</html>
```
Comment (chances) [inline views/projects.ejs]: Revert this change.

**Comment by chances** [comment]
The test file is testing more than one unit. Separate the file into its constituent parts, one for the projects route (`test/routes/projects.js`) and one for the projects module (`test/lib/projects.js`).

Also, the projects module test cases should be structured using nested describes, the outer describing the module and inner describes for the module's functions. (As in the [auth module tests](https://github.com/RickyV33/Jama-Software/blob/LogInPage/test/lib/auth.js).)

File: test/lib/projects.js [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    {},
+    {},
+    {}
+  ]
+];
+
+let ajv = new Ajv();
+let validate = ajv.compile(projectListSchema);
+
+describe('Schema Validation for projects.js', function () {
```
Comment (ibilal) [inline test/lib/projects.js]: Maybe simplify the describe titles to a specific spec, function, class, get/post request etc. Instead of a complete sentence. Nested within an outer describe for the module

File: test/lib/projects.js [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  // Array 0 is for checking against completely valid JSON
+  // Array 1 is for checking against JSON with an invalid entry in the first set
+  let filterJSON = [
+    [
+      {id: 234, name: 'Chatty'},
+      {id: 345, name: 'IIBA BABOK'},
+      {id: 456, name: 'Integrated System'}
+    ], [
+      {id: 345, name: 'IIBA BABOK'},
+      {id: 456, name: 'Integrated System'}
```
Comment (ibilal) [inline test/lib/projects.js]: is this entry in the array supposed to be invalid? It looks the same as element 0

File: test/lib/projects.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    let parsed = projects.parseProjectList(mockProjects);
+    console.log(mockProjects);
+    let result = validate(parsed);
+    expect(parsed).to.deep.equal(filterJSON[0]);
+    expect(result).to.be.true();
+  });
+
+  it('Should return a valid JSON with two of three projects when name is null in the first project', function () {
+    // When id is valid and name is invalid
+    mockProjects[0].fields.name = null;
```
Comment (ibilal) [inline test/lib/projects.js]: Couldn't you create the array with all of these test cases and then write a test for each index into the array. Instead of setting it to valid values and then manually changing the values in each test.

**Comment by ibilal** [comment]
besides my comment about the describes and manually modifying json test case array, LGTM as Chance would say

File: test/lib/projects.js [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+describe('Schema Validation for projects.js', function () {
+  projectFixtureCases.forEach(function (fixture) {
+    let result = validate(fixture);
+    if (result === true) {
+      it('Should return true when the JSON schema object is VALID\n', function () {
+        expect(result).to.be.true();
+      });
+    } else {
+      it('JSON schema object should be INVALID with the credentials\n \tname: ' + fixture[0].name + ' and ID: ' + fixture[0].id + '\n', function () {
+        // console.log('\tShould be invalid with name: ' + fixture[0].name + ' and ID: ' + fixture[0].id);
```
Comment (chances) [inline test/lib/projects.js]: Remove this commented log

**Comment by RickyV33** [comment]
LGTM
